src/v/cloud_storage/tests/segment_meta_cstore_test.cc

```diff
@@ -0,0 +1,776 @@
+}
+
+BOOST_AUTO_TEST_CASE(test_segment_meta_cstore_frame_append_tx_xor) {
+    delta_xor_frame frame(initial_xor);
+    append_tx_test_case(100000, frame);
+}
+
+BOOST_AUTO_TEST_CASE(test_segment_meta_cstore_frame_append_tx_delta) {
+    delta_delta_frame frame(initial_delta);
+    append_tx_test_case(100000, frame);
+}
+
+BOOST_AUTO_TEST_CASE(test_segment_meta_cstore_col_append_tx_xor) {
+    delta_xor_column col(initial_xor);
+    append_tx_test_case(100000, col);
+}
+
+BOOST_AUTO_TEST_CASE(test_segment_meta_cstore_col_append_tx_delta) {
+    delta_delta_column col(initial_delta);
+    append_tx_test_case(100000, col);
+}
+
+template<class column_t>
+void iter_test_case(const int64_t num_elements, column_t& column) {
+    size_t total_size = 0;
+    std::vector<int64_t> expected;
+    int64_t ix = 0;
+    for (int64_t i = 0; i < num_elements; i++) {
+        ix += random_generators::get_int(1, 100);
+        column.append(ix);
+        expected.push_back(ix);
+        total_size++;
+    }
+    BOOST_REQUIRE_EQUAL(total_size, column.size());
+
+    int i = 0;
+    for (auto it = column.begin(); it != column.end(); ++it) {
+        BOOST_REQUIRE_EQUAL(it.index(), i);
+        BOOST_REQUIRE_EQUAL(*it, expected[i++]);
+    }
+}
+
+BOOST_AUTO_TEST_CASE(test_segment_meta_cstore_frame_iter_xor) {
+    delta_xor_frame frame(initial_xor);
+    iter_test_case(100000, frame);
+}
+
+BOOST_AUTO_TEST_CASE(test_segment_meta_cstore_frame_iter_delta) {
+    delta_delta_frame frame(initial_delta);
+    iter_test_case(100000, frame);
+}
+
+BOOST_AUTO_TEST_CASE(test_segment_meta_cstore_col_iter_xor) {
+    delta_xor_column col(initial_xor);
+    iter_test_case(100000, col);
+}
+
+BOOST_AUTO_TEST_CASE(test_segment_meta_cstore_col_iter_delta) {
+    delta_delta_column col(initial_delta);
+    iter_test_case(100000, col);
+}
+
+template<class column_t>
+void find_test_case(const int64_t num_elements, column_t& column) {
+    size_t total_size = 0;
+    std::vector<int64_t> samples;
+    int64_t ix = 0;
+    for (auto i = 0; i < num_elements; i++) {
+        ix += random_generators::get_int(1, 100);
+        column.append(ix);
+        if (samples.empty() || random_generators::get_int(10) == 0) {
+            samples.push_back(ix);
+        }
+        total_size++;
+    }
+    BOOST_REQUIRE_EQUAL(total_size, column.size());
+    std::random_device rd;
+    std::mt19937 g(rd());
+    std::shuffle(samples.begin(), samples.end(), g);
+
+    for (auto expected : samples) {
+        auto it = column.find(expected);
+        BOOST_REQUIRE(it != column.end());
+        auto actual = *it;
+        BOOST_REQUIRE_EQUAL(actual, expected);
+    }
+}
+
+BOOST_AUTO_TEST_CASE(test_segment_meta_cstore_frame_find_xor) {
+    delta_xor_frame frame(initial_xor);
+    find_test_case(100000, frame);
+}
+
+BOOST_AUTO_TEST_CASE(test_segment_meta_cstore_frame_find_xor_small) {
+    delta_xor_frame frame(initial_xor);
+    find_test_case(random_generators::get_int(1, 16), frame);
```

Comment:
I see a large amount of random number use in these tests: especially in situations like this test, it's potentially painful.  If there was an issue with num_elements=1 containers, then this test would pass 15/16 times.

I know we have existing tests that do similar stuff, but...

Perhaps now is a good time to introduce explicit seeding of random number generators in tests.  I don't think this has to be too heavyweight, if we introduce some environment variable for the seed, make a helper for intializing a RNG based on the environment seed, then have tests like this construct a RNG instance instead of relying on the thread-local default instance.

Reply:
I think it's better to do this separately. This will affect a lot of other tests. Looks like a pretty big change.